sql,colexec: add support for wrapping LocalPlanNode

yuzefovich opened the pull request:
**tree: unwrap DOidWrapper before casting**

When performing a cast from `DOidWrapper`, we're interested in casting
the wrapped datum, so this commit adds the unwrapping to the top of
`performCast`.

Release note: None

**colexec: introduce streaming mode to the columnarizer**

Previously, the columnarizer would always buffer up tuples (dynamically,
up to `coldata.BatchSize()`) before emitting them as output; however,
there are several processors which are of "streaming" nature and such
buffering approach is not compatible when wrapping them. This commit
introduces a streaming mode of operation to the columnarizer which is
used when wrapping a streaming processor (newly introduced marker
interface that currently is implemented by the `planNodeToRowSource` and
the changefeed processors).

Note that there are still some unresolved issues around the changefeed
processors, so they aren't being wrapped into the vectorized flows.

Release note: None

**colexec: add support for wrapping LocalPlanNode**

This commit adds the support for wrapping `LocalPlanNode` core. It also
changes the meaning of `experimental_always` to allow wrapping of
JoinReader and LocalPlanNode cores. The latter allows us to remove
VectorizeAlwaysException.

Notable change is that now the materializer created when wrapping
row-execution processors are no longer added to the set of releasables
at the flow cleanup because in some cases it could be released before
being closed. Namely, this would occur if we have a subquery with
LocalPlanNode core and a materializer is added in order to wrap that
core - what will happen is that all releasables are put back into their
pools upon the subquery's flow cleanup, yet the subquery planNode tree
isn't closed yet since its closure is down when the main planNode tree
is being closed.

Another change is fixing the tracing in the cFetcher during mutations
(we were accessing public only columns whereas we should have been
accessing deletable ones - that's what row.Fetcher does in two places
that needed a fix).

Addresses: #57268.

Release note: None

yuzefovich committed:
colexec: introduce streaming mode to the columnarizer

Previously, the columnarizer would always buffer up tuples (dynamically,
up to `coldata.BatchSize()`) before emitting them as output; however,
there are several processors which are of "streaming" nature and such
buffering approach is not compatible when wrapping them. This commit
introduces a streaming mode of operation to the columnarizer which is
used when wrapping a streaming processor (newly introduced marker
interface that currently is implemented by the `planNodeToRowSource` and
the changefeed processors).

Note that there are still some unresolved issues around the changefeed
processors, so they aren't being wrapped into the vectorized flows.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/55909)
<!-- Reviewable:end -->

yuzefovich commented:
This prototype shows the performance different of **-4.43%**:
- `vectorize_row_count_threshold=1000`
```
Highest sequence written: 295698. Can be passed as --write-seq=R295698 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        5067380         8445.6      0.8      0.9      1.5      2.4     92.3  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         265936          443.2      2.1      2.0      3.4      5.8    130.0  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        5333316         8888.9      0.9      0.9      1.9      2.8    130.0  
```
- `vectorize_row_count_threshold=0`
```
Highest sequence written: 284305. Can be passed as --write-seq=R284305 to the next run.

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0        4841248         8068.7      0.9      0.9      1.6      2.5     46.1  read

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__total
  600.0s        0         255233          425.4      2.1      2.0      3.4      5.8     18.9  write

_elapsed___errors_____ops(total)___ops/sec(cum)__avg(ms)__p50(ms)__p95(ms)__p99(ms)_pMax(ms)__result
  600.0s        0        5096481         8494.1      0.9      0.9      2.0      3.0     46.1  
```

yuzefovich committed:
colexec: add support for wrapping LocalPlanNode

This commit adds the support for wrapping `LocalPlanNode` core. It also
changes the meaning of `experimental_always` to allow wrapping of
JoinReader and LocalPlanNode cores. The latter allows us to remove
VectorizeAlwaysException.

Notable change is that now the materializer created when wrapping
row-execution processors are no longer added to the set of releasables
at the flow cleanup because in some cases it could be released before
being closed. Namely, this would occur if we have a subquery with
LocalPlanNode core and a materializer is added in order to wrap that
core - what will happen is that all releasables are put back into their
pools upon the subquery's flow cleanup, yet the subquery planNode tree
isn't closed yet since its closure is down when the main planNode tree
is being closed.

Another change is fixing the tracing in the cFetcher during mutations
(we were accessing public only columns whereas we should have been
accessing deletable ones - that's what row.Fetcher does in two places
that needed a fix).

Release note: None

yuzefovich committed:
tree: unwrap DOidWrapper before casting

When performing a cast from `DOidWrapper`, we're interested in casting
the wrapped datum, so this commit adds the unwrapping to the top of
`performCast`.

Release note: None

yuzefovich commented:
I believe I flushed out all of the failures that I've observed on this branch, so I think it is RFAL.

@jordanlewis I currently separated out the fourth commit that I'd like you to take a look at (and I'll squash it into the third one). The issue showed up in this [build](https://teamcity.cockroachdb.com/viewLog.html?buildId=2483817&buildTypeId=Cockroach_UnitTests) on `orms` logic test where we have an array of `DOidWrappers` that we want to cast to an array of strings with different Oids. I'm not sure whether the current approach will work in all cases, but it seems reasonable to me.

asubiotto commented:
Is this RFAL?

jordanlewis commented on pkg/sql/sem/tree/casts.go:
```diff
@@ -808,6 +808,11 @@ func performCastWithoutPrecisionTruncation(ctx *EvalContext, d Datum, t *types.T
+		if w, ok := d.(*DOidWrapper); ok {
+			// If we're casting a DOidWrapper, then we want to cast the wrapped
+			// datum. It is also reasonable to lose the old Oid value too.
+			d = w.Wrapped
```
I think you want to use the `UnwrapDatum` method here, which will unwrap a whole stack of DOidWrappers if there is one.

jordanlewis replied:
Then again, we should probably just eagerly do that up top.

yuzefovich commented:
Several new failures seem to have popped since the threshold has been lowered to 0, and I'll need to look into those.

I extracted the first two commits into a separate PR #57644 which is RFAL.

yuzefovich commented:
I now believe that many flakes in that build are caused by the randomized batch size cluster setting propagation, so I rebased on top of the metamorphic build PR, and let's see what happens.

yuzefovich commented:
The build now seems to be green (apart from a single backup flake), so I think it is RFAL. Note that only the last 3 commits belong to this PR.

asubiotto reviewed (changes requested):
> Notable change is that now the materializer created when wrapping
> row-execution processors are no longer added to the set of releasables
> at the flow cleanup because in some cases it could be released before
> being closed.

Is this a problem? What are the effects of this? Is this different from `Closer`s?

> Reviewed 57 of 57 files at r5, 1 of 1 files at r6, 13 of 13 files at r7, 8 of 8 files at r8, 9 of 9 files at r9, 9 of 9 files at r10, 1 of 1 files at r11, 11 of 11 files at r12, 48 of 48 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-:-MO_zXa6CQSas8n_J4hn:b-sx3w84)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/conn_executor_internal_test.go, line 259 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MOa0-l19uOJKmCBEnm8:-MOa0-l19uOJKmCBEnm9:br9uq0x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d81626552412e5437f286cce587bf029e38d8f5c/pkg/sql/conn_executor_internal_test.go#L259)):*
> ```Go
> 	distSQLMetrics := execinfra.MakeDistSQLMetrics(time.Hour /* histogramWindow */)
> 	gw := gossip.MakeOptionalGossip(nil)
> 	tempEngine, tempFS, err := storage.NewTempEngine(ctx, base.DefaultTestTempStorageConfig(st), base.DefaultTestStoreSpec)
> ```

Is this an in-mem engine? See `storage.NewDefaultInMem` if not. Also, why did you have to add this?
___
*[pkg/sql/colexec/columnarizer.go, line 37 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MO_woPG5dOpA91urcwc:-MO_woPG5dOpA91urcwd:b-x76eit) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b3b5b04a1198f74d217e21f7c2f181aa5469f7/pkg/sql/colexec/columnarizer.go#L37)):*
> ```Go
> 	// phase.
> 	ColumnarizerDefaultMode ColumnarizerMode = iota
> 	// ColumnarizerStreamingMode is the mode of operation in which the
> ```

Why does making the columnarizer implement the streaming mode solve any problems? Isn't it an issue if we, say, have another operator reading from the columnarizer and it's a buffering processor? Or is it ok for those types of flows to exist since the equivalent row execution flow will do the same.
___
*[pkg/sql/colexec/columnarizer.go, line 66 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MO_yCO78EzOpFXIMcfK:-MO_yCO8CsDGGLxaWFW2:b-k6n3gy) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b3b5b04a1198f74d217e21f7c2f181aa5469f7/pkg/sql/colexec/columnarizer.go#L66)):*
> ```Go
> 
> // NewColumnarizer returns a new Columnarizer.
> func NewColumnarizer(
> ```

Might be nicer to "hide" this`mode` option to avoid having to change always pass in a mode if we mostly want default behavior:
```
NewBufferingColumnarizer(opts...) {
    return newColumnarizer(opts, DefaultMode)
}

NewStreamingColumnarizer(opts...) {
    return newColumnarizer(opts, StreamingMode)
}

newColumnarizer(opts..., mode mode) {

}
```

Also, maybe we should rename `ColumnarizerDetaultMode` to `ColumnarizerBufferingMode` since it's more descriptive? (i.e. it avoids the question of what is the default mode?)


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
>> Notable change is that now the materializer created when wrapping
row-execution processors are no longer added to the set of releasables
at the flow cleanup because in some cases it could be released before
being closed.

> Is this a problem? What are the effects of this? Is this different from `Closers`?

I don't think it is a problem - the code that closes the `planNode` tree is a bit broken in a sense that we cannot close the subquery's tree until after the main query runs, and releasing the materializer that wraps a LocalPlanNode is the only - I think - special case that outlives the flow that it was created by (as a reminder, we're putting all of the objects into their pools on the flow shutdown). So what happens is that subquery's flow is cleaned up yet the `planNode`s end up holding onto those materializers (via other adapters - `planNodeToRowSource`).

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-:-MObsSIX0Z0WqBu7j5dA:b-njjejt)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/conn_executor_internal_test.go, line 259 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MOa0-l19uOJKmCBEnm8:-MOcGdpwBuHcSc49YWiu:b-fo3mj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/d81626552412e5437f286cce587bf029e38d8f5c/pkg/sql/conn_executor_internal_test.go#L259)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Is this an in-mem engine? See `storage.NewDefaultInMem` if not. Also, why did you have to add this?
</blockquote></details>

`TestPortalsDestroyedOnTxnFinish` fails with `error setting up flow: FS unset on DiskQueueCfg`, so we actually need to set the file system for all vectorized flows given that in `vectorizedFlow.Setup` we always create a `DiskQueueCfg`.
___
*[pkg/sql/colexec/columnarizer.go, line 37 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MO_woPG5dOpA91urcwc:-MOcA1Qq8AKWcrYLA1xK:bi5na73) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b3b5b04a1198f74d217e21f7c2f181aa5469f7/pkg/sql/colexec/columnarizer.go#L37)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Why does making the columnarizer implement the streaming mode solve any problems? Isn't it an issue if we, say, have another operator reading from the columnarizer and it's a buffering processor? Or is it ok for those types of flows to exist since the equivalent row execution flow will do the same.
</blockquote></details>

I think there might be only one very special `planNode` that requires this streaming behavior - `hookFnNode`. In its `Next` method it waits for the context cancellation or an error for shutting down or for the next row to be sent on the channel. The buffering behavior of the columnarizer can block indefinitely long - for example, in case of the changefeeds.

In `TestChangefeedBasics/sinkless` if we make the columnarizer buffering, we observe the following situation: there are 5 total changes on the table, so 5 events are emitted; the columnarizer puts the 1st into the first batch, then 2nd and 3rd into the second batch, then 4th and 5th into the third batch of 4 capacity, and the columnarizer asks for more rows, but they never arrive.

At the moment, I believe making a "local" decision whether to make the columnarizer streaming or not works in all cases - AFAIK these `hookFnNode`s are the only things that could block indefinitely when `Next` is called, so we simply wanna make sure to not buffer anything that is coming out of them. I'm guessing that changefeeds might be the only special case, so we could customize the columnarizer to be streaming only on `hookFnNode`s or even more specifically on changefeed `hookFnNode`, but I think it's worth being on the safer side and make all `LocalPlanNode` cores wrapped with the streaming columnarizer.

I cannot say that I have 100% confidence in this approach, but I haven't encountered any other special cases apart from the changefeeds. I'm assuming that if there is something that I'm missing, it'll be exposed by the nightly and unit tests, especially since we are now randomizing `coldata.BatchSize`.

To test my theory about changefeed flow being the only special case, I'll open up a separate PR and see whether anything else fails.
___
*[pkg/sql/colexec/columnarizer.go, line 66 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MO_yCO78EzOpFXIMcfK:-MObyChc2i9U4HoeKJrh:b-1bh1n9) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b3b5b04a1198f74d217e21f7c2f181aa5469f7/pkg/sql/colexec/columnarizer.go#L66)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

Might be nicer to "hide" this`mode` option to avoid having to change always pass in a mode if we mostly want default behavior:
```
NewBufferingColumnarizer(opts...) {
    return newColumnarizer(opts, DefaultMode)
}

NewStreamingColumnarizer(opts...) {
    return newColumnarizer(opts, StreamingMode)
}

newColumnarizer(opts..., mode mode) {

}
```

Also, maybe we should rename `ColumnarizerDetaultMode` to `ColumnarizerBufferingMode` since it's more descriptive? (i.e. it avoids the question of what is the default mode?)
</blockquote></details>

I initially also had "buffering" and "streaming" modes, but I like this suggestion. Done.
___
*[pkg/sql/sem/tree/casts.go, line 814 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MNaOhI1gbCI9gfWrDw7-r4-814:-MObyJgF3jxZa3zyJRk5:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ca7f6575033072d209376438e6c878ff85a64f0d/pkg/sql/sem/tree/casts.go#L814)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Then again, we should probably just eagerly do that up top.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55909)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/columnarizer.go, line 37 at r12](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MO_woPG5dOpA91urcwc:-MOdb9qJ1jj2A9je7MmW:b-ttyiwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/96b3b5b04a1198f74d217e21f7c2f181aa5469f7/pkg/sql/colexec/columnarizer.go#L37)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I think there might be only one very special `planNode` that requires this streaming behavior - `hookFnNode`. In its `Next` method it waits for the context cancellation or an error for shutting down or for the next row to be sent on the channel. The buffering behavior of the columnarizer can block indefinitely long - for example, in case of the changefeeds.

In `TestChangefeedBasics/sinkless` if we make the columnarizer streaming, we observe the following situation: there are 5 total changes on the table, so 5 events are emitted; the columnarizer puts the 1st into the first batch, then 2nd and 3rd into the second batch, then 4th and 5th into the third batch of 4 capacity, and the columnarizer asks for more rows, but they never arrive.

At the moment, I believe making a "local" decision whether to make the columnarizer streaming or not works in all cases - AFAIK these `hookFnNode`s are the only things that could block indefinitely when `Next` is called, so we simply wanna make sure to not buffer anything that is coming out of them. I'm guessing that changefeeds might be the only special case, so we could customize the columnarizer to be streaming only on `hookFnNode`s or even more specifically on changefeed `hookFnNode`, but I think it's worth being on the safer side and make all `LocalPlanNode` cores wrapped with the streaming columnarizer.

I cannot say that I have 100% confidence in this approach, but I haven't encountered any other special cases apart from the changefeeds. I'm assuming that if there is something that I'm missing, it'll be exposed by the nightly and unit tests, especially since we are now randomizing `coldata.BatchSize`.

To test my theory about changefeed flow being the only special case, I'll open up a separate PR and see whether anything else fails.
</blockquote></details>

Looks like my hypothesis might be correct since this [build](https://teamcity.cockroachdb.com/viewLog.html?buildId=2522763&) on a testing PR has only a couple of expected flakes (some linter issues and a benchmark data that needs to be updated in this PR), so I have a bit more confidence now in the current approach (of "local" decision on the mode of the columnarizer).


<!-- Sent from Reviewable.io -->

yuzefovich commented:
I believe I fleshed out all remaining places where the update was needed, so it is RFAL.

asubiotto reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 57 of 57 files at r14, 11 of 11 files at r15, 1 of 54 files at r16.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-:-MOfAIAy-3Jaq8DvSzbA:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @yuzefovich)

___
*[pkg/sql/conn_executor_internal_test.go, line 259 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MOa0-l19uOJKmCBEnm8:-MOf8cEA4iRlU6uLUiX6:b1waw5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/d81626552412e5437f286cce587bf029e38d8f5c/pkg/sql/conn_executor_internal_test.go#L259)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`TestPortalsDestroyedOnTxnFinish` fails with `error setting up flow: FS unset on DiskQueueCfg`, so we actually need to set the file system for all vectorized flows given that in `vectorizedFlow.Setup` we always create a `DiskQueueCfg`.
</blockquote></details>

:+1: what about the in-mem part?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-:-MOg4kfA7J2YEPch8oGm:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/conn_executor_internal_test.go, line 259 at r13](https://reviewable.io/reviews/cockroachdb/cockroach/55909#-MOa0-l19uOJKmCBEnm8:-MOg4rcm0uRu3djSZRDS:b-u08pic) ([raw file](https://github.com/cockroachdb/cockroach/blob/d81626552412e5437f286cce587bf029e38d8f5c/pkg/sql/conn_executor_internal_test.go#L259)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

:+1: what about the in-mem part?
</blockquote></details>

The current invocation creates an in-memory engine and FS. Using `NewDefaultInMem` is not sufficient since it only returns the engine and not the FS, but we're only interested in the FS.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
New logic test needs to be updated.

bors r-

yuzefovich commented:
bors r+
